x/stakeibc/keeper/callbacks.go

```diff
@@ -166,3 +171,137 @@ func WithdrawalBalanceCallback(k Keeper, ctx sdk.Context, args []byte, query icq
+
+	// get tokens using the validator's conversion rate
+	for i, v := range zone.Validators {
+		k.Logger(ctx).Info(fmt.Sprintf("DELCB %s", v.Address))
+		if v.Address == qdel.ValidatorAddress {
+			delAmtInt64, err := cast.ToInt64E(v.DelegationAmt)
+			if err != nil {
+				k.Logger(ctx).Error(fmt.Sprintf("unable to convert delegationAmt to uint64, err: %s", err.Error()))
+				return err
+			}
+
+			// convert shares to tokens using the exchange rate
+			// TODO: make sure conversion math precision matches the sdk's staking module's version (we did it slightly differently)
+			// note: truncateInt per https://github.com/cosmos/cosmos-sdk/blob/cb31043d35bad90c4daa923bb109f38fd092feda/x/staking/types/validator.go#L431
+			qNumTokens := qdel.Shares.Mul(v.TokensFromShares).TruncateInt()
+			k.Logger(ctx).Info(fmt.Sprintf("DelegationCallback: zone %s validator %s prevNtokens %v qNumTokens %v", zone.ChainId, v.Address, v.DelegationAmt, qNumTokens))
+			if qNumTokens.Uint64() < v.DelegationAmt {
+				// TODO(later) add some safety checks here (e.g. we could query the slashing module to confirm the decr in tokens was due to slash)
+				// update our records of the total stakedbal and of the validator's delegation amt
+				// NOTE:  we assume any decrease in delegation amt that's not tracked via records is a slash
+				slashAmt := v.DelegationAmt - qNumTokens.Uint64()
+				slashAmtInt64, err := cast.ToInt64E(slashAmt)
+				if err != nil {
+					k.Logger(ctx).Error(fmt.Sprintf("unable to convert slashAmt to uint64, err: %s", err.Error()))
+					return err
+				}
+				weightInt64, err := cast.ToInt64E(v.Weight)
+				if err != nil {
+					k.Logger(ctx).Error(fmt.Sprintf("unable to convert weight to uint64, err: %s", err.Error()))
+					return err
+				}
+
+				slashPct := sdk.NewDec(slashAmtInt64).Quo(sdk.NewDec(delAmtInt64))
+				k.Logger(ctx).Info(fmt.Sprintf("ICQ'd delAmt mismatch zone %s validator %s delegator %s records was %d icq shows %d slashAmt %d slashPct %d... UPDATING!",
+					zone.ChainId, v.Address, qdel.DelegatorAddress, v.DelegationAmt, qNumTokens, slashAmt, slashPct))
+				// TODO (not priority) move rate limiting logic to new rate limiting module
+				if slashPct.GT(sdk.NewDec(10).Quo(sdk.NewDec(100))) {
+					k.Logger(ctx).Error(fmt.Sprintf("DELCB | slashed but ABORTING bc slash GT10pct: query shows slash of %v", slashPct))
+					return sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "DELCB | slashed but ABORTING bc slash GT10pct: query shows slash of %v", slashPct)
+				}
+				// slash the validator's weight
+				weightMul := sdk.NewDec(qNumTokens.Int64()).Quo(sdk.NewDec(delAmtInt64))
+
+				zone.StakedBal -= slashAmtInt64
+				v.DelegationAmt -= slashAmt
+				v.Weight = sdk.NewDec(weightInt64).Mul(weightMul).TruncateInt().Uint64()
+			}
+			// write back to state and break (reset TokensFromShares for clarity, so we're not tempted to use it again later)
+			v.TokensFromShares = sdk.NewDec(0)
```

Comment:
Can we rename `TokensFromShares` -> `TokensFromSharesExchRate`? I was initially confused by this variable because it sounds like it's storing a number of tokens.

It feels a bit risky to only reset this to 0 at the end of the daisy-chained callback, I know that right now the only place we use this is in this code path, but if the daisy chain were to fail halfway through, the validator could have stale `TokensFromShares` across epochs. I see two possible ways to resolve this (curious if you can think of others)
1. At the beginning of each epoch, iterate all validators on all host zones and set `TokensFromShares` to 0 (slightly inefficient perhaps, but the safest approach)
2. Restructure this as follows
```
message ValidatorExchangeRate {
    string tokensFromSharesExchRate  = 1 [
        (gogoproto.customtype) = "github.com/cosmos/cosmos-sdk/types.Dec",
        (gogoproto.nullable)   = false
    ];
    uint64 epochNumber = 2;

}
```
then add a function
```
func (k Keeper) GetInternalValidatorExchRate(v Validator) sdk.Dec {
    // check that the exchange rate on the validator is scoped to the current epoch
    // return it
}
```

Reply:
Scoping the internal exchange rate to an epochNumber (your suggestion #2) might solve this issue entirely. If we check the epochNumber on an internalExchRate before using it, why do we need to reset to 0 at the beginning the epoch (suggestion #1) ?

Will rename the variable, I agree it's confusing (borrowed from cosmos-sdk distr module nomenclature fwiw, but that doesn't mean we shouldn't improve it)

Reply:
I don't think we need to do both, either 1 or 2 should suffice

Reply:
followed approach 1, done